scripts/format-queries.lua

```diff
@@ -0,0 +1,216 @@
+          lines[#lines] = lines[#lines] .. ' '
+        end
+      end
+    until true
+  end
+end
+
+---@param bufnr integer
+---@param queries string
+local function format(bufnr, queries)
+  local lines = { '' }
+  local map = {
+    ['format.ignore'] = {}, -- Ignore the node and its children
+    ['format.indent.begin'] = {}, -- +1 shiftwidth for all nodes after this
+    ['format.indent.end'] = {}, -- -1 shiftwidth for all nodes after this
+    ['format.indent.dedent'] = {}, -- -n (default 1) shiftwidth for this line only
+    ['format.prepend-space'] = {},
+    ['format.prepend-newline'] = {},
+    ['format.append-space'] = {},
+    ['format.append-newline'] = {},
+    ['format.cancel-append'] = {}, -- Some cases we need to not append newline (eof, final pattern)
+  }
+  local root = ts.get_parser(bufnr, 'query'):parse(true)[1]:root()
+  local query = ts.query.parse('query', queries)
+  for id, node, metadata in query:iter_captures(root, bufnr) do
+    if query.captures[id]:sub(1, 1) ~= '_' then
+      map[query.captures[id]][node:id()] = metadata or {}
+    end
+  end
+
+  iter(bufnr, root, lines, map, 0)
+  vim.api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
+end
+
+-- Query to control the formatter
+local format_queries = [[
+;;query
+;; Making sure all top-level nodes are separated
+(program
+  (_) @format.append-newline)
+(program
+  (_) @format.cancel-append .)
+(program
+  (_) @_comment
+  .
+  (_) @format.prepend-newline
+  (#not-has-type? @_comment comment))
+[
+  ":"
+  "."
+] @format.append-space
+(
+  "." @format.prepend-space @format.cancel-append
+  .
+  ")")
+(list
+  "[" @format.indent.begin
+  .
+  (_)
+  .
+  (_)
+  "]" @format.indent.dedent)
+(list
+  "[" @format.append-space
+  .
+  (_)
+  .
+  "]" @format.prepend-space)
+(list
+  (capture) @format.prepend-space)
+(list
+  (_) @format.append-newline
+  (#not-has-type? @format.append-newline capture))
+(named_node
+  [
+    "_"
+    name: (identifier)
+  ] @format.indent.begin
+  [
+    (list)
+    (grouping)
+    (negated_field)
+    (field_definition)
+    (named_node)
+    (anonymous_node)
+    (predicate)
+    "."
+  ] @format.append-newline)
+(named_node
+  [
+    "_"
+    name: (identifier)
+  ] @format.indent.begin
+  (_) @format.cancel-append
+  .
+  ")")
+(named_node
+  (capture) @format.prepend-space)
+(anonymous_node
+  (capture) @format.prepend-space)
+(predicate
+  (parameters
+    (_) @format.prepend-space))
+(grouping
+  "(" @format.ignore
+  .
+  (_
+    (capture)) @format.indent.begin
+  .
+  (predicate) .)
+(grouping
+  "(" @format.indent.begin
+  [
+    (anonymous_node)
+    (named_node)
+    (list)
+    (predicate)
+    "."
+  ] @format.append-newline
+  (_) .)
+(comment) @format.append-newline
+]]
+
+---@type string[]
+local files = vim.fn.split(vim.fn.glob(root_path .. '/runtime/queries/**/*.*'))
+
+for _, file in ipairs(files) do
+  local buf = vim.fn.bufadd(file)
+  vim.fn.bufload(file)
+  format(buf, format_queries)
+end
+
+vim.cmd('wqa!')
```

Comment:
I don't think that's needed with `-l`, either.

Reply:
`-l` doesn't actually write any file. I don't know exactly how to parse a raw string so I opted for `bufload` -> `bufadd` -> `write`

Reply:
Then of course you have to `vim.cmd.write()`; you just don't need the `quit()`.